{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "outside-dispatch",
   "metadata": {},
   "source": [
    "# Specification\n",
    "\n",
    "- Program name: missingMotif.ipynb\n",
    "- Input/Output: Input filename: 1st required parameter to main(), Output STDOUT\n",
    "\n",
    "## options:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "humanitarian-think",
   "metadata": {},
   "source": [
    "- --minMotif  minimum motif size to evaluate (int)\n",
    "- --maxMotif maximum motif size to evaluate (int)\n",
    "- --cutoff  Z-score cutoff (float)\n",
    "\n",
    "Ex: main(\"input.fa\",[ \"--minMotif=3\", \"--maxMotif=8\", \"--cutoff=-4.\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "reliable-directive",
   "metadata": {},
   "source": [
    "# Background\n",
    "Many prokaryotic genomes encode Restriction-Modification systems (RM) that act to protect the cell from invading DNA by cutting at specific sites (frequently short 4-6 base reverse complement palindromes). They also protect host DNA from being cut by modifying any sites within the host DNA that would be sites for the cutting enzyme. It is not unusual to find that these enzymes are adjacent to each other in the host genome. (see: https://en.wikipedia.org/wiki/Restriction_modification_system).\n",
    "\n",
    "Meselson and Stahl (1958) showed that replication of the bacterial dsDNA genome happens in a semi-conservative fashion (see: https://en.wikipedia.org/wiki/Prokaryotic_DNA_replication). The resulting circular chromosome(s) would then be transiently hemi-methylated because the new daughter strand may not been have exposed to the protective methylation of the RM system before its cutting activity arrives. This creates a race between RM cutting and protecting activities may expose the cell to strand breaks.\n",
    "\n",
    "Is a selective advantage incurred for the cell by allowing mutations to accumulate at these sites? If so, then would we expect the host genome to become depleted of RM sites over time? Can we use this depletion process as a signature for the existence of the specific RM systems present in the genome? \n",
    "\n",
    "How can we find sequences that are missing in a genome?\n",
    "\n",
    "# Assignment\n",
    "\n",
    "Using this notebook, write a proper style BME205 solution that reads a fasta file and ranks motifs based on how statistically underrepresented the specific motif is. We need only consider sequence motifs that are up to about 8 in length ( this is a program option). In order to use the equation that we developed in class, the minimum size should be at least 3 ( this is an option also). We will find that there are quite a few of these, so we need to specify a statistical cutoff. For this assignment, we will use Z-scores. The scores we are looking for will be negative since we are looking for underrepresented sequences (this cutoff is a program option too). Finally, we need to sort the output by motif size and then z-score within size groups and print it out.\n",
    "\n",
    "Remember that sequences that we get from DNA sequencing are from only one strand of DNA. This means that when we see a AAGGTT, this implies that AACCTT (the reverse complement) is present on the other strand. We should count these sequences as equivalent. \n",
    "\n",
    "Your report should be sorted by Zscore within motif-Size, such that the longest motifs are at the beginning of the report and, within motifs of a given size, print the most significant motifs first. Print the k-mer pairs in alpha order. ( AAA:TTT and not TTT:AAA)\n",
    "\n",
    "The extra credit for this assignment is particularly interesting. If you have time, this will present better results. \n",
    "\n",
    "Here is an example (non extra credit):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "absolute-prompt",
   "metadata": {},
   "outputs": [],
   "source": [
    "# sequence: reverse     count      Expect Zscore\n",
    "# AAAATTTA:TAAATTTT\t835\t1737.66\t-21.66\n",
    "# GATTAATA:TATTAATC\t550\t1326.89\t-21.33\n",
    "# AATTAATA:TATTAATT\t929\t1839.72\t-21.24\n",
    "# AATTAATC:GATTAATT\t977\t1861.79\t-20.51\n",
    "# ATAATTAA:TTAATTAT\t1033\t1926.49\t-20.36\n",
    "# GAAATTTA:TAAATTTC\t378\t1031.07\t-20.34\n",
    "# CCGATCGC:GCGATCGG\t1323\t2284.74\t-20.12\n",
    "# GCGATCGA:TCGATCGC\t1221\t2121.78\t-19.56\n",
    "# ACGATCGC:GCGATCGT\t1188\t2035.31\t-18.78\n",
    "# GTTTAAAA:TTTTAAAC\t519\t1151.17\t-18.63\n",
    "# AAATATTA:TAATATTT\t738\t1444.79\t-18.60\n",
    "# AAAATTTG:CAAATTTT\t857\t1591.95\t-18.42\n",
    "# CATTAATC:GATTAATG\t479\t1037.49\t-17.34\n",
    "# ATATATAA:TTATATAT\t271\t736.19\t-17.15\n",
    "# ATTTAAAG:CTTTAAAT\t419\t941.16\t-17.02\n",
    "# AAATATTG:CAATATTT\t687\t1287.74\t-16.74\n",
    "# CAAATTTA:TAAATTTG\t670\t1265.24\t-16.74\n",
    "# CTTTAAAC:GTTTAAAG\t468\t995.13\t-16.71"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "medieval-running",
   "metadata": {},
   "source": [
    "# Hints\n",
    "- For **genomes that include multiple fasta records, do NOT concatenate them**. They should be reported as a single genome, though we do not want to create false kmers that would result from concatenation\n",
    "- **Avoid zero divide errors!** Close examination of our null model will show that if we have a zero in the denominator, we will have a zero in the numerator. If you have pre-populated your count dictionary, then you will need to handle dividing by zero.\n",
    "- Only consider kmers that are constructed from the **canonical bases {A, C, G, T}**.\n",
    "- Z-score is calculated as: $\\frac{s-\\mu}{sd}$, where s is the specific count you are converting, and mean and standard deviation are given by µ and sd. Both µ and sd describe the null distribution. In our case, we can treat the expected distribution as a binomial where: $\\mu=np,sd=\\sqrt{np(1-p)}$. n = genome size (approximately), and p = probability of success, given the null distribution.\n",
    "- c(x) defined as the count of x, generally the count of some substring. For example, for kmers of size 4: $K=k_{1}k_{2}k_{3}k_{4}, \n",
    "Pr(K)=\\frac{c(K)}{N},\n",
    "E(K) = \\frac{c(k_{1}k_{2}k_{3})\\cdot c(k_{2}k_{3}k_{4})}{c(k_{2}k_{3})}$\n",
    "(see below for the derivation of our null model) We can then calculate n and p, and derive µ and sd. \n",
    "- To keep your scores lined up, use tabs to separate values in each line. The following format string will be of use: \n",
    "\n",
    "print('{0:8}:{1:8}\\t{2:0d}\\t{3:0.2f}\\t{4:0.2f}'.format(seq, rSeq, count,E,pVal))\n",
    "\n",
    "\n",
    "# Null Model derivation\n",
    "If we assume complete independence of the any of the characters that make up a sequence K, we could approximate the expected count of K, E(K),  as:\n",
    "$\\begin{matrix} \n",
    "K=k_{1}k_{2}k_{3}k_{4}, Pr(k_{i})=\\frac{c(k_{i})}{N} \\\\ \n",
    "E(K)=N * Pr(K) = N * \n",
    "Pr(k_{1})\n",
    "Pr(k_{2})\n",
    "Pr(k_{3})\n",
    "Pr(k_{4}) \n",
    "\\end{matrix} \n",
    "$\n",
    "\n",
    "the above is a markov(0) approximation for the expected count (E) of our sequence K.\n",
    "\n",
    "Rarely is this assumption of independence a good null model, due to many factors including codon bias which drives selection of 3-mers based on underlying amino acid frequencies. We can model this dependence using a markov model, where the probability associated with any symbol is conditioned on the preceding characters. For a kmer of size 4, we can condition our character probabilities using the preceding two bases without requiring use of the count of the specific 4-mer that we are trying to approximate(see note on information content). We need to consider edge conditions in our approximation. In general, we can use a model of order k-2, where k is the size of the kmer as long as we have sufficient data, therefore:\n",
    "\n",
    "$\n",
    "\\begin{matrix} \n",
    "Pr(K)=\\frac{E(K)}{N} = \n",
    "Pr(k_{1})\n",
    "Pr(k_{2}|k_{1})\n",
    "Pr(k_{3}|k_{1}k_{2})\n",
    "Pr(k_{4}|k_{2}k_{3}) \\\\\n",
    "= \\frac{c(k_{1})}{N}\n",
    "\\frac{c(k_{1}k_{2})}{c(k_{1})}\n",
    "\\frac{c(k_{1} k_{2} k_{3})} {c(k_{1} k_{2})}\n",
    "\\frac{c(k_{2} k_{3} k_{4})} {c(k_{2} k_{3})} \\\\\n",
    "= \\frac{1}{N}\n",
    "\\frac{c(k_{1} k_{2} k_{3})c(k_{2} k_{3} k_{4})} {c(k_{2} k_{3})}\n",
    "\\end{matrix}\n",
    "$\n",
    "\n",
    "note: the conditional probability Pr(a | b) is read as the probability of finding sequence a, given sequence b. In this case we treat sequence b as adjacent and left of a, and ask what is the probability of finding sequence a, given that b is immediately to the left of a. This means that we should count the occurrences of string ba, among all strings that start with b (taken as the count of all sequence b).\n",
    "\n",
    "note on information content: What would happen if we were trying to predict the frequencies of 4-mers, and our null model contained the frequency of 4-mers? \n",
    "\n",
    "2021 Note: Make sure to review the module video on this. The derivation is a bit more elaborate.\n",
    "\n",
    "# Extra Credit\n",
    "Among all kmers of size k, we have both the measured value and the expected value for that specific kMer. We are interested in knowing which of  these sequences produce counts that are surprisingly small. If we were to normalize our actual counts by the expected count, we can then look to the smallest values,(<1)  as those that evoke the greatest surprise. Throughout this assignment, we will refer to the actual, integer counts of a kMer as \"counts\". When we normalize an actual count by its expected value, we will refer to those as \"normalized counts\".\n",
    "We can then calculate a mean and standard deviation of those normalized counts within size-groups(sequence groups of the same size, for example, all of the 8-mers), to gain a quantitative measure of that surprise. We can think of this as the distribution of surprise within each size-group. kMers that yeild a normalized count near 1 are not very surprising, because the actual and expected count values are close to each other.\n",
    "\n",
    "For a possible 5 extra points:\n",
    "\n",
    "- add option **--kScoring** which evaluates to true only when the flag is specified by the user\n",
    "- calculate Z-scores over the normalized score found for each kMer, within each size-group. \n",
    "- use this Z-score for score cutoff\n",
    "- include a column for the p-value associated with this Z-score. \n",
    "\n",
    "p-Values can be calculated from the CDF of the binomial or Normal distribution. Feel free to use scipy for this. scipy.stats.norm.cdf(z). For those of you using an M1 Mac, make sure to use Anaconda\n",
    "\n",
    "# Notes\n",
    "- I considered adding pseudo counts to this assignment but have deferred that to the next (4-Oct)\n",
    "- Added the full derivation of the null model (4-Oct)\n",
    "- clarified the use of E - the expected value, and in our case, an expected motif count. (4-Oct)\n",
    "- clarified the use of the count function c(x) (4-Oct)\n",
    "- clarified the use of counts of strings in conditional probabilities. (4-Oct)\n",
    "- clarified the avoidance of zero divide or the use of kmers that contain non-canonical bases (4-Oct)\n",
    "- **Extra Credit** To calculate a distribution over normalized counts for kmers in a particular size-group, we need to calculate the mean and sd of that size-group. This can be done by finding a normalized count of a kmer of a given size, then computing a running sum of those counts of that kmer-size, a running sum of the square of those normalized counts of that kmer-size, and the number of kmers we found of each kmer-size. These three statistics would need to be tabulated for each of the kmer-sizes that we are working with. So...lets think of these as three vectors that each contain kMax values. The three vectors contain kMax elements: $N_k, {\\sum{normalizedCounts}}_k, {\\sum{normalizedCounts^2}}_k$.\n",
    "\n",
    "In python terms, we need:\n",
    "- statsN[1..kMax], \n",
    "- statsSum[1..kMax], and \n",
    "- statsSum2[1..kMax]. \n",
    "\n",
    "Each of these vectors (lists) correspond to a tabulation N = the number of kmers of some specific size(k), the sum of the the normalized-counts of specific size(k), and sum2 = the sum of the squared normalized-counts of specific size(k). \n",
    "Now, we can calculate mean and SD for the group of kmers of size k by:\n",
    "\n",
    "$ \\mu_{k}=\\frac{sum_k}{N_k}$\n",
    "\n",
    "$ SD_{k} =\\sqrt{\\frac{sum2_k}{N_k}-\\frac{(sum_{k})^{2}}{N_{k}^2}}$\n",
    "\n",
    "## 2018 additions\n",
    "\n",
    "- binned the reverse complement sequences. This produces better Evalues and a more concise report. ( 1-Oct)\n",
    "\n",
    "## 2019 additions\n",
    "\n",
    "- added calculation of normalized kMer counts with significance determined by  their distribution within groups of size k.\n",
    "\n",
    "## 2020 addition\n",
    "\n",
    "- its good to think about our kmer counting by asking how many names does this kmer have? For example, the kmer AAAA when viewed from the opposite strand is TTTT. This bit of data has 2 names, so when we are counting we make sure that counts of this kmer by either of its names goes into the same bin. How should we handle a reverse complement palindrome like TATA ? When viewed from either strand, it has the same name.\n",
    "\n",
    "## 2021 addition:\n",
    "\n",
    "- we are using notebooks now to formalize inspections and promote Python learning. Changed the expected input.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "korean-brick",
   "metadata": {},
   "source": [
    "## Inspection Intro\n",
    "Provide design level information for your inspection team here.Examples:\n",
    "- How did you count kmers of each size without having to read the data multiple times?\n",
    "- How did you organize the computing of mean and SD for each of the kmer groups ?\n",
    "- How did you compute Z-scores for each kMer in a kMer group?\n",
    "- How did you sort the final report to include most significant Z-scores at the top of each kmer-size group?\n",
    "- How did you organize the program ?\n",
    "\n",
    "## Inspection Results\n",
    "- What did your inspection team find? Which of the inspection results did you use, how did you resolve them, and which did you decide against? Who was on your inspection team?\n",
    "\n",
    "# Code that may be useful"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "differential-championship",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "\n",
    "class FastAreader:\n",
    "    def __init__(self, fname=''):\n",
    "        '''contructor: saves attribute fname '''\n",
    "\n",
    "        self.fname = fname\n",
    "        self.fileH = None\n",
    "\n",
    "    def doOpen(self):\n",
    "        if self.fname == '':\n",
    "            return sys.stdin\n",
    "        else:\n",
    "            return open(self.fname)\n",
    "\n",
    "    def readFasta(self):\n",
    "\n",
    "        header = ''\n",
    "        sequence = ''\n",
    "\n",
    "        with self.doOpen() as self.fileH:\n",
    "\n",
    "            header = ''\n",
    "            sequence = ''\n",
    "\n",
    "            # skip to first fasta header\n",
    "            line = self.fileH.readline()\n",
    "            while not line.startswith('>'):\n",
    "                line = self.fileH.readline()\n",
    "            header = line[1:].rstrip()\n",
    "\n",
    "            for line in self.fileH:\n",
    "                if line.startswith('>'):\n",
    "                    yield header, sequence\n",
    "                    header = line[1:].rstrip()\n",
    "                    sequence = ''\n",
    "                else:\n",
    "                    sequence += ''.join(line.rstrip().split()).upper()\n",
    "\n",
    "        yield header, sequence\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "czech-irrigation",
   "metadata": {},
   "outputs": [],
   "source": [
    "class CommandLine():\n",
    "    '''\n",
    "    Handle the command line, usage and help requests.\n",
    "\n",
    "    CommandLine uses argparse, now standard in 2.7 and beyond. \n",
    "    it implements a standard command line argument parser with various argument options,\n",
    "    a standard usage and help, and an error termination mechanism do-usage_and_die.\n",
    "\n",
    "    attributes:\n",
    "    all arguments received from the commandline using .add_argument will be\n",
    "    avalable within the .args attribute of object instantiated from CommandLine.\n",
    "    For example, if myCommandLine is an object of the class, and requiredbool was\n",
    "    set as an option using add_argument, then myCommandLine.args.requiredbool will\n",
    "    name that option.\n",
    "\n",
    "    '''\n",
    "\n",
    "    def __init__(self, inOpts=None):\n",
    "        '''\n",
    "        CommandLine constructor.\n",
    "        Implements a parser to interpret the command line argv string using argparse.\n",
    "        '''\n",
    "\n",
    "        import argparse\n",
    "        self.parser = argparse.ArgumentParser(\n",
    "            description='Program prolog - a brief description of what this thing does',\n",
    "            epilog='Program epilog - some other stuff you feel compelled to say',\n",
    "            add_help=True,  # default is True\n",
    "            prefix_chars='-',\n",
    "            usage='%(prog)s [options] -option1[default] <input >output'\n",
    "        )\n",
    "\n",
    "        self.parser.add_argument('-l', '--minMotif', nargs='?', default=1, action='store',\n",
    "                                 help='min kMer size ')\n",
    "        self.parser.add_argument('-m', '--maxMotif', nargs='?', default=8, action='store',\n",
    "                                 help='max kMer size ')\n",
    "        self.parser.add_argument('-c', '--cutoff', nargs='?', type=float, default=.01, action='store',\n",
    "                                 help='Zscore cutoff')\n",
    "\n",
    "\n",
    "        self.parser.add_argument('-v', '--version', action='version', version='%(prog)s 0.1')\n",
    "        if inOpts is None:\n",
    "            self.args = self.parser.parse_args()\n",
    "        else:\n",
    "            self.args = self.parser.parse_args(inOpts)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fuzzy-rendering",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Genome:\n",
    "    \"\"\" \"\"\"\n",
    "    def __init__(self, minK, maxK):\n",
    "        \"\"\"Construct the object to count and analyze kMer frequencies\"\"\"\n",
    "        self.minK = int(minK)\n",
    "        self.maxK = int(maxK) \n",
    "    \n",
    "    # get rc\n",
    "    def getrc(self, s):\n",
    "        n = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A'}\n",
    "        return \"\".join(n[i] for i in reversed(s))\n",
    "        \n",
    "    # calculate counts and expected\n",
    "    def count(self, dna):\n",
    "        \n",
    "        import math\n",
    "        \n",
    "        N = len(dna)\n",
    "        res = {}\n",
    "        p1 = {} # store 1 preceding bases according to k\n",
    "        p2 = {} # store 2 preceding bases\n",
    "        for i in range(self.minK, self.maxK + 1):\n",
    "            srs = []\n",
    "            srs1 = []\n",
    "            srs2 = []\n",
    "            for j in range(0, len(dna) - i + 1):               \n",
    "                s = dna[j:j+i]\n",
    "                sr = self.getrc(s)\n",
    "                # fill res\n",
    "                if s not in res and s not in srs:\n",
    "                    srs.append(sr)\n",
    "                    res[s] = [1, None, None]\n",
    "                elif s in res: \n",
    "                    res[s][0] += 1   \n",
    "                elif s in srs and s != sr: \n",
    "                    res[sr][0] += 1   \n",
    "                # fill p1  \n",
    "                s1 = dna[j:j+i-1]\n",
    "                sr1 = self.getrc(s1)\n",
    "                if s1 not in p1 and s1 not in srs1:\n",
    "                    srs1.append(sr1)\n",
    "                    p1[s1] = 1\n",
    "                elif s1 in p1: \n",
    "                    p1[s1] += 1  \n",
    "                elif s1 in srs1 and s1 != sr1: \n",
    "                    p1[sr1] += 1   \n",
    "                # fill p2    \n",
    "                s2 = dna[j:j+i-2]\n",
    "                sr2 = self.getrc(s2)\n",
    "                if s2 not in p2 and s2 not in srs2:\n",
    "                    srs2.append(sr2)\n",
    "                    p2[s2] =1\n",
    "                elif s2 in p2:\n",
    "                    p2[s2] += 1\n",
    "                elif s2 in srs2 and s2 != sr2:\n",
    "                    p2[sr2] += 1\n",
    "                              \n",
    "            # calculate expected number\n",
    "            keys = list(res.keys())\n",
    "            for key in keys:\n",
    "                if len(key) == i:\n",
    "                    n1 = key[0:len(key)-1]\n",
    "                    n1r = self.getrc(n1)\n",
    "                    n2 = key[1:len(key)]\n",
    "                    n2r = self.getrc(n2)\n",
    "                    d = key[1:len(key)-1]\n",
    "                    dr = self.getrc(d)\n",
    "                    if n1 in p1 and n2 in p1 and d in p2:\n",
    "                        exp = (p1[n1] * p1[n2]) / p2[d]  \n",
    "                    elif n1 not in p1 and n2 in p1 and d in p2:\n",
    "                        exp = (p1[n1r] * p1[n2]) / p2[d]\n",
    "                    elif n1 in p1 and n2 not in p1 and d in p2:\n",
    "                        exp = (p1[n1] * p1[n2r]) / p2[d]\n",
    "                    elif n1 in p1 and n2 in p1 and d not in p2: \n",
    "                        exp = (p1[n1] * p1[n2]) / p2[dr] \n",
    "                    elif n1 not in p1 and n2 not in p1 and d in p2:\n",
    "                        exp = (p1[n1r] * p1[n2r]) / p2[d] \n",
    "                    elif n1 in p1 and n2 not in p1 and d not in p2:    \n",
    "                        exp = (p1[n1] * p1[n2r]) / p2[dr] \n",
    "                    elif n1 not in p1 and n2 in p1 and d not in p2:\n",
    "                        exp = (p1[n1r] * p1[n2]) / p2[dr]\n",
    "                    elif n1 not in p1 and n2 not in p1 and d not in p2:\n",
    "                        exp = (p1[n1r] * p1[n2r]) / p2[dr]         \n",
    "                    res[key][1] = exp\n",
    "                    \n",
    "                    # calculate z\n",
    "                    p = exp / N\n",
    "                    sd = math.sqrt(exp*(1-p))\n",
    "                    z = (res[key][0]-exp) / sd\n",
    "                    res[key][2] = z\n",
    "                    \n",
    "            # reset helper dictionary\n",
    "            p1.clear()\n",
    "            p2.clear()  \n",
    "        del p1\n",
    "        del p2   \n",
    "        \n",
    "        return res  # dict[key] : (count, expected, None)\n",
    "    \n",
    "    def modifier(self, res_dict, cut):\n",
    "        ks = list(res_dict.keys())\n",
    "        for k in ks:\n",
    "            if float(res_dict[k][2]) > float(cut):\n",
    "                del res_dict[k]  \n",
    "        return res_dict\n",
    "    \n",
    "    def sort(self, dict, min, max):\n",
    "        new = {}\n",
    "        for i in range(int(max), int(min)-1, -1):\n",
    "            temp = {} # store temp different length base\n",
    "            for key in list(dict.keys()):\n",
    "                if len(key) == i:\n",
    "                    temp[key] = dict[key]\n",
    "            # sort\n",
    "            tot_v = temp.values()\n",
    "            z_v = []\n",
    "            for i in tot_v:\n",
    "                z_v.append(i[2]) # z score  \n",
    "            for v in sorted(z_v):\n",
    "                for k in list(temp.keys()):\n",
    "                    if temp[k][2] == v:\n",
    "                        new[k] = temp[k]\n",
    "        return new            \n",
    "                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "buried-airline",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32mc:\\Users\\YuTia\\OneDrive\\Desktop\\BME\\BME205\\WK1\\missingMotif.ipynb Cell 10\u001b[0m in \u001b[0;36m<cell line: 32>\u001b[1;34m()\u001b[0m\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/YuTia/OneDrive/Desktop/BME/BME205/WK1/missingMotif.ipynb#X12sZmlsZQ%3D%3D?line=28'>29</a>\u001b[0m             fout\u001b[39m.\u001b[39mwrite(\u001b[39m'\u001b[39m\u001b[39m{0:8}\u001b[39;00m\u001b[39m:\u001b[39m\u001b[39m{1:8}\u001b[39;00m\u001b[39m\\t\u001b[39;00m\u001b[39m{2:0d}\u001b[39;00m\u001b[39m\\t\u001b[39;00m\u001b[39m{3:0.2f}\u001b[39;00m\u001b[39m\\t\u001b[39;00m\u001b[39m{4:0.2f}\u001b[39;00m\u001b[39m'\u001b[39m\u001b[39m.\u001b[39mformat(\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/YuTia/OneDrive/Desktop/BME/BME205/WK1/missingMotif.ipynb#X12sZmlsZQ%3D%3D?line=29'>30</a>\u001b[0m                 seq, rSeq, count, E, pVal) \u001b[39m+\u001b[39m \u001b[39m'\u001b[39m\u001b[39m\\n\u001b[39;00m\u001b[39m'\u001b[39m)    \n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/YuTia/OneDrive/Desktop/BME/BME205/WK1/missingMotif.ipynb#X12sZmlsZQ%3D%3D?line=31'>32</a>\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39m__name__\u001b[39m \u001b[39m==\u001b[39m \u001b[39m\"\u001b[39m\u001b[39m__main__\u001b[39m\u001b[39m\"\u001b[39m:\n\u001b[1;32m---> <a href='vscode-notebook-cell:/c%3A/Users/YuTia/OneDrive/Desktop/BME/BME205/WK1/missingMotif.ipynb#X12sZmlsZQ%3D%3D?line=32'>33</a>\u001b[0m     main(inFile\u001b[39m=\u001b[39;49m\u001b[39m\"\u001b[39;49m\u001b[39mZm4-genomic.fna\u001b[39;49m\u001b[39m\"\u001b[39;49m,options\u001b[39m=\u001b[39;49m[ \u001b[39m\"\u001b[39;49m\u001b[39m--minMotif=3\u001b[39;49m\u001b[39m\"\u001b[39;49m, \u001b[39m\"\u001b[39;49m\u001b[39m--maxMotif=8\u001b[39;49m\u001b[39m\"\u001b[39;49m, \u001b[39m\"\u001b[39;49m\u001b[39m--cutoff=-5.\u001b[39;49m\u001b[39m\"\u001b[39;49m])\n",
      "\u001b[1;32mc:\\Users\\YuTia\\OneDrive\\Desktop\\BME\\BME205\\WK1\\missingMotif.ipynb Cell 10\u001b[0m in \u001b[0;36mmain\u001b[1;34m(inFile, options)\u001b[0m\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/YuTia/OneDrive/Desktop/BME/BME205/WK1/missingMotif.ipynb#X12sZmlsZQ%3D%3D?line=8'>9</a>\u001b[0m f\u001b[39m.\u001b[39mwrite(dna)\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/YuTia/OneDrive/Desktop/BME/BME205/WK1/missingMotif.ipynb#X12sZmlsZQ%3D%3D?line=10'>11</a>\u001b[0m thisGenome \u001b[39m=\u001b[39m Genome(cl\u001b[39m.\u001b[39margs\u001b[39m.\u001b[39mminMotif, \n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/YuTia/OneDrive/Desktop/BME/BME205/WK1/missingMotif.ipynb#X12sZmlsZQ%3D%3D?line=11'>12</a>\u001b[0m                     cl\u001b[39m.\u001b[39margs\u001b[39m.\u001b[39mmaxMotif) \u001b[39m# setup a Genome object\u001b[39;00m\n\u001b[1;32m---> <a href='vscode-notebook-cell:/c%3A/Users/YuTia/OneDrive/Desktop/BME/BME205/WK1/missingMotif.ipynb#X12sZmlsZQ%3D%3D?line=12'>13</a>\u001b[0m \u001b[39mdict\u001b[39m \u001b[39m=\u001b[39m thisGenome\u001b[39m.\u001b[39;49mcount(dna)\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/YuTia/OneDrive/Desktop/BME/BME205/WK1/missingMotif.ipynb#X12sZmlsZQ%3D%3D?line=13'>14</a>\u001b[0m \u001b[39mdict\u001b[39m \u001b[39m=\u001b[39m thisGenome\u001b[39m.\u001b[39mmodifier(\u001b[39mdict\u001b[39m, cl\u001b[39m.\u001b[39margs\u001b[39m.\u001b[39mcutoff)\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/YuTia/OneDrive/Desktop/BME/BME205/WK1/missingMotif.ipynb#X12sZmlsZQ%3D%3D?line=14'>15</a>\u001b[0m \u001b[39mdict\u001b[39m \u001b[39m=\u001b[39m thisGenome\u001b[39m.\u001b[39msort(\u001b[39mdict\u001b[39m, cl\u001b[39m.\u001b[39margs\u001b[39m.\u001b[39mminMotif, cl\u001b[39m.\u001b[39margs\u001b[39m.\u001b[39mmaxMotif)\n",
      "\u001b[1;32mc:\\Users\\YuTia\\OneDrive\\Desktop\\BME\\BME205\\WK1\\missingMotif.ipynb Cell 10\u001b[0m in \u001b[0;36mGenome.count\u001b[1;34m(self, dna)\u001b[0m\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/YuTia/OneDrive/Desktop/BME/BME205/WK1/missingMotif.ipynb#X12sZmlsZQ%3D%3D?line=25'>26</a>\u001b[0m \u001b[39mfor\u001b[39;00m j \u001b[39min\u001b[39;00m \u001b[39mrange\u001b[39m(\u001b[39m0\u001b[39m, \u001b[39mlen\u001b[39m(dna) \u001b[39m-\u001b[39m i \u001b[39m+\u001b[39m \u001b[39m1\u001b[39m):               \n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/YuTia/OneDrive/Desktop/BME/BME205/WK1/missingMotif.ipynb#X12sZmlsZQ%3D%3D?line=26'>27</a>\u001b[0m     s \u001b[39m=\u001b[39m dna[j:j\u001b[39m+\u001b[39mi]\n\u001b[1;32m---> <a href='vscode-notebook-cell:/c%3A/Users/YuTia/OneDrive/Desktop/BME/BME205/WK1/missingMotif.ipynb#X12sZmlsZQ%3D%3D?line=27'>28</a>\u001b[0m     sr \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mgetrc(s)\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/YuTia/OneDrive/Desktop/BME/BME205/WK1/missingMotif.ipynb#X12sZmlsZQ%3D%3D?line=28'>29</a>\u001b[0m     \u001b[39m# fill res\u001b[39;00m\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/YuTia/OneDrive/Desktop/BME/BME205/WK1/missingMotif.ipynb#X12sZmlsZQ%3D%3D?line=29'>30</a>\u001b[0m     \u001b[39mif\u001b[39;00m s \u001b[39mnot\u001b[39;00m \u001b[39min\u001b[39;00m res \u001b[39mand\u001b[39;00m s \u001b[39mnot\u001b[39;00m \u001b[39min\u001b[39;00m srs:\n",
      "\u001b[1;32mc:\\Users\\YuTia\\OneDrive\\Desktop\\BME\\BME205\\WK1\\missingMotif.ipynb Cell 10\u001b[0m in \u001b[0;36mGenome.getrc\u001b[1;34m(self, s)\u001b[0m\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/YuTia/OneDrive/Desktop/BME/BME205/WK1/missingMotif.ipynb#X12sZmlsZQ%3D%3D?line=8'>9</a>\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mgetrc\u001b[39m(\u001b[39mself\u001b[39m, s):\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/YuTia/OneDrive/Desktop/BME/BME205/WK1/missingMotif.ipynb#X12sZmlsZQ%3D%3D?line=9'>10</a>\u001b[0m     n \u001b[39m=\u001b[39m {\u001b[39m'\u001b[39m\u001b[39mA\u001b[39m\u001b[39m'\u001b[39m: \u001b[39m'\u001b[39m\u001b[39mT\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mC\u001b[39m\u001b[39m'\u001b[39m: \u001b[39m'\u001b[39m\u001b[39mG\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mG\u001b[39m\u001b[39m'\u001b[39m: \u001b[39m'\u001b[39m\u001b[39mC\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mT\u001b[39m\u001b[39m'\u001b[39m: \u001b[39m'\u001b[39m\u001b[39mA\u001b[39m\u001b[39m'\u001b[39m}\n\u001b[1;32m---> <a href='vscode-notebook-cell:/c%3A/Users/YuTia/OneDrive/Desktop/BME/BME205/WK1/missingMotif.ipynb#X12sZmlsZQ%3D%3D?line=10'>11</a>\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39m\"\u001b[39;49m\u001b[39m\"\u001b[39;49m\u001b[39m.\u001b[39;49mjoin(n[i] \u001b[39mfor\u001b[39;49;00m i \u001b[39min\u001b[39;49;00m \u001b[39mreversed\u001b[39;49m(s))\n",
      "\u001b[1;32mc:\\Users\\YuTia\\OneDrive\\Desktop\\BME\\BME205\\WK1\\missingMotif.ipynb Cell 10\u001b[0m in \u001b[0;36m<genexpr>\u001b[1;34m(.0)\u001b[0m\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/YuTia/OneDrive/Desktop/BME/BME205/WK1/missingMotif.ipynb#X12sZmlsZQ%3D%3D?line=8'>9</a>\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mgetrc\u001b[39m(\u001b[39mself\u001b[39m, s):\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/YuTia/OneDrive/Desktop/BME/BME205/WK1/missingMotif.ipynb#X12sZmlsZQ%3D%3D?line=9'>10</a>\u001b[0m     n \u001b[39m=\u001b[39m {\u001b[39m'\u001b[39m\u001b[39mA\u001b[39m\u001b[39m'\u001b[39m: \u001b[39m'\u001b[39m\u001b[39mT\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mC\u001b[39m\u001b[39m'\u001b[39m: \u001b[39m'\u001b[39m\u001b[39mG\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mG\u001b[39m\u001b[39m'\u001b[39m: \u001b[39m'\u001b[39m\u001b[39mC\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mT\u001b[39m\u001b[39m'\u001b[39m: \u001b[39m'\u001b[39m\u001b[39mA\u001b[39m\u001b[39m'\u001b[39m}\n\u001b[1;32m---> <a href='vscode-notebook-cell:/c%3A/Users/YuTia/OneDrive/Desktop/BME/BME205/WK1/missingMotif.ipynb#X12sZmlsZQ%3D%3D?line=10'>11</a>\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39m\"\u001b[39m\u001b[39m\"\u001b[39m\u001b[39m.\u001b[39mjoin(n[i] \u001b[39mfor\u001b[39;00m i \u001b[39min\u001b[39;00m \u001b[39mreversed\u001b[39m(s))\n",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "def main (inFile=None, options = None):\n",
    "    ''' Setup necessary objects, read data and print the final report.'''\n",
    "    cl = CommandLine(options) # setup the command line\n",
    "    sourceReader = FastAreader(inFile) # setup the Fasta reader Object\n",
    "    sourceReader.doOpen()\n",
    "    \n",
    "    dna = (next(sourceReader.readFasta())[1]) # get dna strand\n",
    "        \n",
    "    thisGenome = Genome(cl.args.minMotif, \n",
    "                        cl.args.maxMotif) # setup a Genome object\n",
    "    dict = thisGenome.count(dna)\n",
    "    dict = thisGenome.modifier(dict, cl.args.cutoff)\n",
    "    dict = thisGenome.sort(dict, cl.args.minMotif, cl.args.maxMotif)\n",
    "    \n",
    "    # file output\n",
    "    fout = open(\"myArthrospira.out\", 'w')\n",
    "    fout.write('N = ' + str(len(dna)) + '\\n')\n",
    "    for seq in list(dict.keys()):\n",
    "        rSeq = thisGenome.getrc(seq)\n",
    "        count = dict[seq][0]\n",
    "        E = dict[seq][1]\n",
    "        pVal = dict[seq][2]\n",
    "        if rSeq > seq:\n",
    "            fout.write('{0:8}:{1:8}\\t{2:0d}\\t{3:0.2f}\\t{4:0.2f}'.format(\n",
    "                rSeq, seq, count, E, pVal) + '\\n')\n",
    "        else:\n",
    "            fout.write('{0:8}:{1:8}\\t{2:0d}\\t{3:0.2f}\\t{4:0.2f}'.format(\n",
    "                seq, rSeq, count, E, pVal) + '\\n')    \n",
    "    \n",
    "if __name__ == \"__main__\":\n",
    "    main(inFile=\"Arthrospira-platensis-NIES-39.fna\",options=[ \"--minMotif=3\", \"--maxMotif=8\", \"--cutoff=-5.\"])\n",
    "   "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.7 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "nteract": {
   "version": "0.28.0"
  },
  "vscode": {
   "interpreter": {
    "hash": "899d556501ff83904803b41a96a9ace85150701b625ae9b324dbd777656e52c4"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
