{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TIANGE YU (Jill)\n",
    "## Inspection intro\n",
    "\n",
    "- What data structure did you use to hold your sequence motifs (DNA)?         \n",
    "I am planning to use list of string to hold motifs \n",
    "\n",
    "- Is your profile  structure based on counts or frequencies?            \n",
    "I think it is more easier to biuld profile with frequencies     \n",
    "\n",
    "- Where did you implement iterations (mostly why did you choose this)?          \n",
    "The iteration should be after each run of random motif search. Cause in that case you can compare each best score of random motif search to aviod program trapped in local maximum.   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Inspection Result    \n",
    "\n",
    "- In terms of profile: we think it is more intuitive to use dictionary to store profile with keys of ATGC. I was initially thinking of storing profile using 2d list, but it turns out that 2d list makes it harder to access the keys and it is not memory efficient.        \n",
    "\n",
    "- I was intially misunderstund the concept of pseudocounts. I thought pseudocounts were only used when certain base has zero probablity, that causes deviation of the profile. Then I realized that pseudocounts is added to all base counts.\n",
    "\n",
    "- It is more useful when profile is in frequencies beacause that makes the calculate of score easier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "class CommandLine():\n",
    "    '''\n",
    "    Handle the command line, usage and help requests.\n",
    "\n",
    "    CommandLine uses argparse, now standard in 2.7 and beyond. \n",
    "    it implements a standard command line argument parser with various argument options,\n",
    "    a standard usage and help, and an error termination mechanism do-usage_and_die.\n",
    "\n",
    "    attributes:\n",
    "    all arguments received from the commandline using .add_argument will be\n",
    "    avalable within the .args attribute of object instantiated from CommandLine.\n",
    "    For example, if myCommandLine is an object of the class, and requiredbool was\n",
    "    set as an option using add_argument, then myCommandLine.args.requiredbool will\n",
    "    name that option.\n",
    "\n",
    "    '''\n",
    "\n",
    "    def __init__(self, inOpts=None):\n",
    "        '''\n",
    "        CommandLine constructor.\n",
    "        Implements a parser to interpret the command line argv string using argparse.\n",
    "        '''\n",
    "\n",
    "        import argparse\n",
    "        self.parser = argparse.ArgumentParser(\n",
    "            description='Program prolog - a brief description of what this thing does',\n",
    "            epilog='Program epilog - some other stuff you feel compelled to say',\n",
    "            add_help=True,  # default is True\n",
    "            prefix_chars='-',\n",
    "            usage='%(prog)s [options] -option1[default] <input >output'\n",
    "        )\n",
    "\n",
    "        self.parser.add_argument('-i', type=int, required=True)\n",
    "        self.parser.add_argument('-k', type=int, required=True)\n",
    "        self.parser.add_argument('-p', type=float, required=True)\n",
    "\n",
    "        self.parser.add_argument('-v', '--version', action='version', version='%(prog)s 0.1')\n",
    "        if inOpts is None:\n",
    "            self.args = self.parser.parse_args()\n",
    "        else:\n",
    "            self.args = self.parser.parse_args(inOpts)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Motif():\n",
    "    '''\n",
    "    generate profiles with given motif, with pseudocounts\n",
    "    input: \n",
    "        Dna a list of dna sequence\n",
    "        pc is integer of pseudocounts\n",
    "        ms is list of strings of initiated motifs\n",
    "    output:\n",
    "        dictionary with A, C, G, T as key\n",
    "    ''' \n",
    "    def profile(self, Dna, pc, ms):\n",
    "        n = 4*pc + len(Dna)\n",
    "        p = {'a':[], 'c':[], 'g':[], 't':[]}\n",
    "        for i in range(0, len(ms[0])):\n",
    "            temp = [n[i] for n in ms]\n",
    "            p['a'].append((temp.count('a')+pc)/n)\n",
    "            p['c'].append((temp.count('c')+pc)/n)\n",
    "            p['g'].append((temp.count('g')+pc)/n)\n",
    "            p['t'].append((temp.count('t')+pc)/n)\n",
    "            temp.clear()    \n",
    "        return p\n",
    "    \n",
    "    '''\n",
    "    get new motif based on given profile\n",
    "    input:\n",
    "        Dna: a list of string of dna sequence\n",
    "        pr: profile in dictionary form\n",
    "        k: integer represent k-mer\n",
    "    output:\n",
    "        motif: list of string\n",
    "    '''\n",
    "    def motif(self, Dna, pr, k):\n",
    "        \n",
    "        from math import prod\n",
    "\n",
    "        motifs = []\n",
    "        for seq in Dna:\n",
    "            ss = []\n",
    "            for x in range(0, len(seq)-k+1):\n",
    "                ss.append(seq[x:x+k])\n",
    "\n",
    "            # calculate probability of each possible motif\n",
    "            prob_temp = []\n",
    "            for i in range(0, len(ss)):\n",
    "                list = []\n",
    "                for j in range(0, k):\n",
    "                    list.append(pr[ss[i][j]][j])\n",
    "                prob_temp.append(prod(list))\n",
    "                \n",
    "            # find motif with max possibility\n",
    "            max_v = max(prob_temp)\n",
    "            max_i = prob_temp.index(max_v)\n",
    "            better_m = ss[max_i]\n",
    "            motifs.append(better_m) \n",
    "        return motifs\n",
    "    \n",
    "    '''\n",
    "    use relative entropy method to get relative distance with null model\n",
    "    input: \n",
    "        pr: dictionary form of profile\n",
    "        Dna: a list of string of sequence\n",
    "        pc: int of ps count\n",
    "        k: int of k-mer length\n",
    "    output:\n",
    "        float of score\n",
    "    '''\n",
    "    def getScore(self, pr, Dna, pc, k):\n",
    "        \n",
    "        from math import log2\n",
    "        \n",
    "        # construct null model  \n",
    "        seq = ''\n",
    "        for s in Dna:\n",
    "            seq += s \n",
    "        null = {} # Q(i)\n",
    "        null['a'] = (seq.count('a')+pc)/(len(seq)+4*pc)\n",
    "        null['c'] = (seq.count('c')+pc)/(len(seq)+4*pc)\n",
    "        null['g'] = (seq.count('g')+pc)/(len(seq)+4*pc)\n",
    "        null['t'] = (seq.count('t')+pc)/(len(seq)+4*pc)\n",
    "        \n",
    "        # experimental model\n",
    "        score = 0\n",
    "        for i in range(0, k):\n",
    "            for key in list(pr.keys()):            \n",
    "                score += (pr[key][i])*log2(pr[key][i]/null[key])\n",
    "        \n",
    "        return score\n",
    "    \n",
    "    def RandomizedMotifSearch(self, Dna, k, pc):\n",
    "        \n",
    "        from random import randint\n",
    "\n",
    "        # initiation: randomly pick motif\n",
    "        motifs = []\n",
    "        for s in Dna:\n",
    "            start = randint(0, len(s)-k)\n",
    "            motifs.append(s[start:start+k])\n",
    "        \n",
    "        bestS = 0\n",
    "        bestM = motifs\n",
    "        while True:\n",
    "            cur_pr = self.profile(Dna, pc, motifs)\n",
    "            motifs = self.motif(Dna, cur_pr, k)\n",
    "            cur_score = self.getScore(self.profile(Dna, pc, motifs), Dna, pc, k)\n",
    "            if cur_score > bestS:\n",
    "                bestS = cur_score\n",
    "                bestM = motifs\n",
    "            else:\n",
    "                return bestM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "p1860Crisprs: \n",
      "8.657886340591507\n",
      "CGAAAAACTTAAA\n",
      "\n",
      "paeCrisprs: \n",
      "5.402943872451918\n",
      "CAAAGAAAAACTT\n",
      "\n",
      "parCrisprs: \n",
      "7.9779836447711565\n",
      "CCAAGAAAAACTT\n",
      "\n",
      "pcaCrisprs: \n",
      "7.783555496977209\n",
      "CAAAGAAAAACTT\n",
      "\n",
      "pisCrisprs: \n",
      "8.177546321992208\n",
      "CCCAAAAGATCCA\n",
      "\n",
      "pogCrisprs: \n",
      "7.550960478104058\n",
      "CCAAAAGAACTGT\n",
      "\n",
      "sum.fa: \n",
      "11.890713001365832\n",
      "GAAAAACTTAAAA\n",
      "\n"
     ]
    }
   ],
   "source": [
    "def main (inFile=None, options = None):\n",
    "    \n",
    "    # initialize command line object\n",
    "    cl = CommandLine(options)\n",
    "    \n",
    "    # read fastA file to get DNA\n",
    "    f = open(inFile, 'r')\n",
    "    dna = []\n",
    "    for r in f.readlines():\n",
    "        if '>' not in r:\n",
    "            dna.append(r.strip())  \n",
    "            \n",
    "    # initialize motif object\n",
    "    motif = Motif()\n",
    "    \n",
    "    # repeat\n",
    "    BM = []\n",
    "    bestScore = 0\n",
    "    for r in range(cl.args.i):\n",
    "        curMotif = motif.RandomizedMotifSearch(dna, cl.args.k, cl.args.p)\n",
    "        curProfile = motif.profile(dna, cl.args.p, curMotif)\n",
    "        curScore = motif.getScore(curProfile, dna, cl.args.p, cl.args.k)\n",
    "        if curScore > bestScore:\n",
    "            bestScore = curScore\n",
    "            BM = curMotif\n",
    "\n",
    "    # find consensus\n",
    "    t_p = motif.profile(dna, cl.args.p, BM)\n",
    "    consensus = ''\n",
    "    for i in range(0, cl.args.k):\n",
    "        temp = []\n",
    "        for k in list(t_p.keys()):\n",
    "            temp.append(t_p[k][i])\n",
    "        max_v = max(temp)\n",
    "        max_i = temp.index(max_v)\n",
    "        if max_i == 0:\n",
    "            consensus += 'A'\n",
    "        elif max_i == 1:\n",
    "            consensus += 'C'\n",
    "        elif max_i == 2:\n",
    "            consensus += 'G'\n",
    "        else:\n",
    "            consensus += 'T'       \n",
    "    \n",
    "    print(inFile + ': ')\n",
    "    print(bestScore)\n",
    "    print(consensus + '\\n')   \n",
    "    \n",
    "       \n",
    "if __name__ == \"__main__\":\n",
    "    main(inFile='p1860Crisprs', options=['-i 1000', '-k 13', '-p 1'])\n",
    "    main(inFile='paeCrisprs', options=['-i 5000', '-k 13', '-p 1'])\n",
    "    main(inFile='parCrisprs', options=['-i 1000', '-k 13', '-p 1'])\n",
    "    main(inFile='pcaCrisprs', options=['-i 1000', '-k 13', '-p 1'])\n",
    "    main(inFile='pisCrisprs', options=['-i 1000', '-k 13', '-p 1'])\n",
    "    main(inFile='pogCrisprs', options=['-i 1000', '-k 13', '-p 1'])\n",
    "    main(inFile='sum.fa', options=['-i 1000', '-k 13', '-p 1'])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.7 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "899d556501ff83904803b41a96a9ace85150701b625ae9b324dbd777656e52c4"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
