{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Peptide:\n",
    "    def __init__(self, Text, AA, codon):\n",
    "        self.Text = Text\n",
    "        self.AA = AA\n",
    "        self.codon = {\n",
    "                    'TTT': 'F', 'TCT': 'S', 'TAT': 'Y', 'TGT': 'C', \n",
    "                    'TTC': 'F', 'TCC': 'S', 'TAC': 'Y', 'TGC': 'C', \n",
    "                    'TTA': 'L', 'TCA': 'S', 'TAA': 'STOP', 'TGA': 'STOP',\n",
    "                    'TTG': 'L', 'TCG': 'S', 'TAG': 'STOP', 'TGG': 'W', \n",
    "                    'ATT': 'I', 'ACT': 'T', 'AAT': 'N', 'AGT': 'S', \n",
    "                    'ATC': 'I', 'ACC': 'T', 'AAC': 'N', 'AGC': 'S', \n",
    "                    'ATA': 'I', 'ACA': 'T', 'AAA': 'K', 'AGA': 'R', \n",
    "                    'ATG': 'M', 'ACG': 'T', 'AAG': 'K', 'AGG': 'R', \n",
    "                    'CTT': 'L', 'CCT': 'P', 'CAT': 'H', 'CGT': 'R', \n",
    "                    'CTC': 'L', 'CCC': 'P', 'CAC': 'H', 'CGC': 'R', \n",
    "                    'CTA': 'L', 'CCA': 'P', 'CAA': 'Q', 'CGA': 'R', \n",
    "                    'CTG': 'L', 'CCG': 'P', 'CAG': 'Q', 'CGG': 'R',\n",
    "                    'GTT': 'V', 'GCT': 'A', 'GAT': 'D', 'GGT': 'G', \n",
    "                    'GTC': 'V', 'GCC': 'A', 'GAC': 'D', 'GGC': 'G',\n",
    "                    'GTA': 'V', 'GCA': 'A', 'GAA': 'E', 'GGA': 'G',\n",
    "                    'GTG': 'V', 'GCG': 'A', 'GAG': 'E', 'GGG': 'G'\n",
    "                    }  \n",
    "    \n",
    "    def reverse_compliment(self, string):\n",
    "        '''Get the reverse compliment of the dna string'''\n",
    "        complement = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A'}\n",
    "        rc = \"\".join(complement.get(base, base) for base in reversed(string))\n",
    "        return rc\n",
    "    \n",
    "    def translate(self, string):\n",
    "        '''Translate the chosen string to peptide'''\n",
    "        k = 3\n",
    "        translation = ''\n",
    "        for i in range(0, len(string)-k+1, k):\n",
    "            code = string[i:i+k]\n",
    "            if self.codon[code] == 'STOP':\n",
    "                break\n",
    "            translation += self.codon[code]\n",
    "        return translation\n",
    "    \n",
    "    def find(self):\n",
    "        '''Find substrings of a genome encoding a given amino acid sequence.\n",
    "        Given: \n",
    "            A DNA string Text and an amino acid string Peptide.\n",
    "        Return: \n",
    "            All substrings of Text encoding Peptide (if any such substrings exist).\n",
    "        '''\n",
    "        res = []\n",
    "        k = len(self.AA) * 3\n",
    "        for i in range(0, len(self.Text)-k+1):\n",
    "            code = self.Text[i:i+k]\n",
    "            trans = self.translate(code)\n",
    "            trans_rv = self.translate(self.reverse_compliment(code))\n",
    "            if trans == self.AA or trans_rv == self.AA:\n",
    "                res.append(code)\n",
    "        return res\n",
    "    \n",
    "def main(inFile = None):\n",
    "    '''\n",
    "    Do the main thing\n",
    "    '''\n",
    "    f = open(inFile, 'r')\n",
    "    Text = f.readline().strip()\n",
    "    AA = f.readline().strip()\n",
    "    peptide = Peptide(Text, AA, codon={})\n",
    "    res = peptide.find()\n",
    "    \n",
    "    with open('p16answer', 'w') as fw:\n",
    "        for s in res:\n",
    "            fw.write(s + '\\n')\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main(inFile = 'rosalind_ba4b.txt') \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Inspection\n",
    "### 1. William Gao:\n",
    "\n",
    "- Very clean and concise code! I like that you are using a sliding window of the peptide sequence length to find the sequences. I did it in a different way but your appraoch is much better\n",
    "- I like that you have `translate()`, `reverse_compliment()`, and `find()` methods for your `Peptide` class in an object-oriented way\n",
    "- Great docstrings as well!\n",
    "    - except the docstring for `main()`\n",
    "\n",
    "### 2. Lisa Biswas:\n",
    "- Code is readible and easy to understand and your use of methods is very effective. \n",
    "- Good docstrings, add in-line comments! \n",
    "\n",
    "### 3. Gia Hagos:\n",
    "- Code looks efficient to the point I feel that I've majorly overthought my own work.\n",
    "- Doc strings are very clear and concise.\n",
    "- Well organized and very easy to read."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.7 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "899d556501ff83904803b41a96a9ace85150701b625ae9b324dbd777656e52c4"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
